### PR TITLE
add image policy documentation

### PR DESCRIPTION
**Pull request by deads2k**
Adds documentation explaining how to use and test the ImagePolicy admission plugin.

@simon3z fyi
@smarterclayton beta or tech preview?  What's the distinction again?

### CODE REVIEW

**Comment by simon3z** [comment]
cc @moolitayer

File: admin_guide/image_policy.adoc [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+:toc: macro
+:toc-title:
+
+toc::[]
+
+[[image-policy-overview]]
+== Overview
+
+You can control which images are allowed to run on your cluster using the ImagePolicy
+admission plugin (currently considered beta).  It allows you to:
```
Comment (moolitayer) [inline admin_guide/image_policy.adoc]: ```It allows you to control```? (otherwise 1. below does not read well)

**Comment by moolitayer** [comment]
@simon3z I will create an upstream doc patch linking to this document.

**Comment by simon3z** [comment]
> @simon3z I will create a ManageIQ upstream doc linking here.

@moolitayer I suppose this will be published somewhere in https://docs.openshift.com/enterprise/latest/welcome/index.html
cc @deads2k

**Comment by deads2k** [comment]
@adellape Is there a tag I should use for 3.3 items?  I've got a few.

File: admin_guide/image_policy.adoc [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+ 1. the source of images: which registries can be used to pull images
+ 2. image resolution: force pods to run with immutable digests to ensure the image doesn't change due to a re-tag
+ 3. docker image label restrictions: force an image to have or not have particular labels
+ 4. image annotation restrictions: force an image in the integration docker registry to have or not have particular annotations
+
+
+== Configuring the ImagePolicy admission plugin
+
+To enable this feature, you'll need to configure the plugin the `master-config.yaml`.
```
Comment (moolitayer) [inline admin_guide/image_policy.adoc]: ```the plugin the``` above

File: admin_guide/image_policy.adoc [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+metadata:
+  generateName: test-pod
+spec:
+  containers:
+  - image: <internal registry IP>:5000/<namespace>/image-policy-check:latest
+    name: first
+```
+
+Create a pod pointing to the internal registry using the imported image.  It should work and
+if you look at the image spec, you should the tag unmodified.
```
Comment (moolitayer) [inline admin_guide/image_policy.adoc]: ```you should see```?

**Comment by deads2k** [comment]
Comments addressed.

@moolitayer @simon3z take note.  We changed the API slightly since 3.3 had not shipped.  I've updated the examples to match.

**Comment by deads2k** [comment]
@ahardin-rh another for 1.3/3.3

File: admin_guide/image_policy.adoc [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    the default is "allow all".  If you have any accept rule, the default is "deny all".
+<3> Indicates which resources to enforce rules upon.  If nothing is specified, the default is "pods"
+<4> Indicates that if this rule matches, the pod should be rejected
+<5> List of annotations to match on the Image object's metadata
+<6> If you aren't able to resolve the image, don't fail the pod
+
+====
+
+
+== Testing the ImagePolicy admission plugin
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: s/admission plugin/Admission Plug-in

File: admin_guide/image_policy.adoc [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+<2> Array of rules to evaluate against incoming resources.  If you only have reject==true rules
+    the default is "allow all".  If you have any accept rule, the default is "deny all".
+<3> Indicates which resources to enforce rules upon.  If nothing is specified, the default is "pods"
+<4> Indicates that if this rule matches, the pod should be rejected
+<5> List of annotations to match on the Image object's metadata
+<6> If you aren't able to resolve the image, don't fail the pod
+
+====
+
+
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: We need to add an anchor just above the heading, so:
[[image-policy-testing-image-policy-admission-plug-in]]

File: admin_guide/image_policy.adoc [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+:toc: macro
+:toc-title:
+
+toc::[]
+
+[[image-policy-overview]]
+== Overview
+
+You can control which images are allowed to run on your cluster using the ImagePolicy
+admission plugin (currently considered beta).  It allows you to control:
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: s/plugin/plug-in

Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: Is beta the same thing as "technology preview"?

File: admin_guide/image_policy.adoc [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+ 1. the source of images: which registries can be used to pull images
+ 2. image resolution: force pods to run with immutable digests to ensure the image doesn't change due to a re-tag
+ 3. docker image label restrictions: force an image to have or not have particular labels
+ 4. image annotation restrictions: force an image in the integration docker registry to have or not have particular annotations
+
+
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: Need an anchor and for the heading to be in title case:

[[image-policy-configuring-the-image-policy-admission-plug-in]]
Configuring the ImagePolicy Admission Plug-in

File: admin_guide/image_policy.adoc [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+== Configuring the ImagePolicy admission plugin
+
+To enable this feature, you'll need to configure the plugin in `master-config.yaml`.
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: s/you'll need to/you need to
s/plugin/plug-in

Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: *_master-config.yaml_*

File: admin_guide/image_policy.adoc [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+Here is an annotated example file.
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: s/file./file:

File: admin_guide/image_policy.adoc [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+----
+
+<1> Try to resolve images to an immutable image digest and update the in the pod
+<2> Array of rules to evaluate against incoming resources.  If you only have reject==true rules
+    the default is "allow all".  If you have any accept rule, the default is "deny all".
+<3> Indicates which resources to enforce rules upon.  If nothing is specified, the default is "pods"
+<4> Indicates that if this rule matches, the pod should be rejected
+<5> List of annotations to match on the Image object's metadata
+<6> If you aren't able to resolve the image, don't fail the pod
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: s/aren't/are not
s/don't/do not

File: admin_guide/image_policy.adoc [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+toc::[]
+
+[[image-policy-overview]]
+== Overview
+
+You can control which images are allowed to run on your cluster using the ImagePolicy
+admission plugin (currently considered beta).  It allows you to control:
+
+ 1. the source of images: which registries can be used to pull images
+ 2. image resolution: force pods to run with immutable digests to ensure the image doesn't change due to a re-tag
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: s/doesn't/does not

File: admin_guide/image_policy.adoc [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        - name: allow-images-from-dockerhub
+          onResources:
+          - resource: pods
+          - resource: builds
+          matchRegistries:
+          - docker.io
+
+----
+
+<1> Try to resolve images to an immutable image digest and update the in the pod
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: missing word after "update the"

File: admin_guide/image_policy.adoc [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+<5> List of annotations to match on the Image object's metadata
+<6> If you aren't able to resolve the image, don't fail the pod
+
+====
+
+
+== Testing the ImagePolicy admission plugin
+
+You can use the `openshift/image-policy-check` to test your configuration.
+For example, you could use the information above and then test like this:
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: Suggest rewording:
For example, use the information above, then test like this:

File: admin_guide/image_policy.adoc [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+```
+oc import-image openshift/image-policy-check:latest --confirm
+
+```
+
+Create a pod using a definition like this.  It should work.
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: suggest rewording / remove "It should work"

Create a pod using a definition file:

File: admin_guide/image_policy.adoc [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+metadata:
+  generateName: test-pod
+spec:
+  containers:
+  - image: different-registry/openshift/image-policy-check:latest
+    name: first
+```
+
+Create a pod pointing to the internal registry using the imported image.  It should work and
+if you look at the image spec, you should see a digest in place of the tag.
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: s/spec/specification

File: admin_guide/image_policy.adoc [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+metadata:
+  generateName: test-pod
+spec:
+  containers:
+  - image: <internal registry IP>:5000/<namespace>/image-policy-check:latest
+    name: first
+```
+
+Create a pod pointing to the internal registry using the imported image.  It should work and
+if you look at the image spec, you should see the tag unmodified.
```
Comment (ahardin-rh) [inline admin_guide/image_policy.adoc]: s/spec/specification

**Comment by ahardin-rh** [comment]
@deads2k Thanks! A few comments from me. 

Should this be included in Dedicated as well? If so, I will adjust my labels accordingly. Thanks again!

**Comment by deads2k** [comment]
> Should this be included in Dedicated as well? If so, I will adjust my labels accordingly. Thanks again!

No, it shouldn't, though it does raise an interesting question about whether and how someone in dedicated would want to control this.  @smarterclayton something for us to consider.

**Comment by deads2k** [comment]
Comments addressed.

File: admin_guide/image_policy.adoc (hunk 6 above)
Comment (deads2k) [inline admin_guide/image_policy.adoc]: > Is beta the same thing as "technology preview"?

Similar: https://github.com/openshift/origin/#alpha-and-unsupported-kubernetes-features .  I think this is beta.

**Comment by ahardin-rh** [comment]
@deads2k Thanks! Can you please give it a squash? :bow:

**Comment by deads2k** [comment]
> @deads2k Thanks! Can you please give it a squash?

done

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../admin_guide/image_policy.adoc#admin-guide-image-policy[Image Policy]
|New topic about controlling which images are allowed to run on your cluster using the ImagePolicy admission plug-in.
%
